bin/trin/src/cli.rs

```diff
@@ -0,0 +1,1141 @@
+                state_mb: self.storage_state,
+            },
+        }
+    }
+}
+
+fn check_trusted_block_root(trusted_root: &str) -> Result<B256, String> {
+    if !trusted_root.starts_with("0x") {
+        return Err("Trusted block root must be prefixed with 0x".to_owned());
+    }
+
+    if trusted_root.len() == 66 {
+        return B256::from_str(trusted_root).map_err(|err| format!("HexError: {err}"));
+    }
+    Err(format!(
+        "Invalid trusted block root length: {}, expected 66 (0x-prefixed 32 byte hexstring)",
+        trusted_root.len()
+    ))
+}
+
+impl fmt::Display for TrinConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let web3_ipc_path_str = self.web3_ipc_path.as_path().display().to_string();
+        let json_rpc_url = match &self.web3_transport {
+            Web3TransportType::HTTP => self.web3_http_address.as_str(),
+            Web3TransportType::IPC => &web3_ipc_path_str[..],
+        };
+
+        f.debug_struct("TrinConfig")
+            .field("subnetworks", &self.portal_subnetworks)
+            .field("storage.total", &self.storage_total)
+            .field("storage.beacon", &self.storage_beacon)
+            .field("storage.history", &self.storage_history)
+            .field("storage.state", &self.storage_state)
+            .field("ephemeral", &self.ephemeral)
+            .field("json_rpc_url", &json_rpc_url)
+            .field("metrics_enabled", &self.enable_metrics_with_url.is_some())
+            .finish()
+    }
+}
+
+#[derive(Subcommand, Debug, Clone, PartialEq)]
+#[allow(clippy::enum_variant_names)]
+pub enum TrinConfigCommands {
+    CreateDashboard(DashboardConfig),
+}
+
+#[derive(Args, Debug, Default, Clone, PartialEq)]
+#[allow(clippy::enum_variant_names)]
+pub struct DashboardConfig {
+    #[arg(default_value = "http://localhost:3000")]
+    pub grafana_address: String,
+
+    #[arg(default_value = "admin")]
+    pub grafana_username: String,
+
+    #[arg(default_value = "admin")]
+    pub grafana_password: String,
+
+    #[arg(default_value = "http://host.docker.internal:9090")]
+    pub prometheus_address: String,
+}
+
+pub fn create_dashboard(
+    dashboard_config: DashboardConfig,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let grafana = GrafanaAPI::new(
+        dashboard_config.grafana_username,
+        dashboard_config.grafana_password,
+        dashboard_config.grafana_address,
+    );
+
+    let prometheus_uid = grafana.create_datasource(
+        "prometheus".to_string(),
+        "prometheus".to_string(),
+        dashboard_config.prometheus_address,
+    )?;
+
+    // Create a dashboard from each pre-defined template
+    for template_path in DASHBOARD_TEMPLATES {
+        let dashboard_url = grafana.create_dashboard(template_path, &prometheus_uid)?;
+        println!("Dashboard successfully created: {dashboard_url}");
+    }
+
+    Ok(())
+}
+
+impl TrinConfig {
+    pub fn portalnet_config(&self, private_key: B256) -> PortalnetConfig {
```

Comment:
Not sure if I like this pattern, I prefer implementing the From trait for `PortalnetConfig` and `RpcConfig`.

```rust
impl From<TrinConfig> for PortalnetConfig {
...
}

impl From<TrinConfig> for RpcConfig {
...
}

```

Reply:
I can do From for `RpcConfig`, but I can't for `impl From<TrinConfig> for PortalnetConfig {` as it requires the `private_key: B256` argument

Reply:
OK, we can then change the function name to `to_portalnet_config` because we are not consuming `self` and will follow the naming conventions.